packages/trace-viewer/src/ui/uiModeView.tsx

```diff
@@ -336,9 +323,26 @@ export const UIModeView: React.FC<{}> = ({
 
   // Watch implementation.
   React.useEffect(() => {
-    if (!testServerConnection)
+    if (!testServerConnection || !teleSuiteUpdater)
       return;
-    const disposable = testServerConnection.onTestFilesChanged(params => {
+    const disposable = testServerConnection.onTestFilesChanged(async params => {
+      commandQueue.current = commandQueue.current.then(async () => {
+        setIsLoading(true);
+        try {
+          const result = await testServerConnection.listTests({ projects: queryParams.projects, locations: queryParams.args, grep: queryParams.grep, grepInvert: queryParams.grepInvert });
+          teleSuiteUpdater.processListReport(result.report);
+        } catch (e) {
+          // eslint-disable-next-line no-console
+          console.log(e);
+        } finally {
+          setIsLoading(false);
+        }
+      });
+      await commandQueue.current;
+
+      const testModel = teleSuiteUpdater.asModel();
+      const testTree = new TestTree('', testModel.rootSuite, testModel.loadErrors, projectFilters, pathSeparator);
```

Comment:
It is unfortunate that we have to create a new `TestTree` here, even though one would be created naturally through `useMemo()`. I wonder whether there is a better pattern to handle such a thing in React?

Reply:
I agree. The big problem with `useMemo` is that it only updates on the next render, which is too late for this. I feel like this component could benefit from separating some of the underlying state from the presentation. I'd like to do that in a separate PR, though.

Reply:
I agree. The big problem with useMemo is that it only updates on the next render, which is too late for this. I feel like this component could benefit from separating some of the underlying state from the presentation. I'd like to do that in a separate PR, though.